harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -20,6 +20,16 @@
     pass
 
 
+def iterate_batches(dataset_len: int, start: int, end: int) -> Iterable[Tuple[int, range]]:
+    """
+    A helper to iterate over a dataset using start and end indices.
+    """
+    for epoch_idx in range(start // dataset_len, end // dataset_len + 1):
```

Comment:
we can't really use `self` considering this being defined as a utility function we could maybe add this to the trial controller but instead, IMO it might be better to name that division to make it more explicit? I update it.

Reply:
There are two issues with this function. One issue is that this function is introducing the dependency of our training loop on the concept of the epoch, which is dependent on the length of the training loader. If the training loader returns a wrong number, the current training loop still works except for the epoch-related callbacks. The other issue is that this function is duplicate logic from `PyTorchTrialController.get_epoch_idx`. I meant to just use `self.context. get_epoch_idx(idx)` directly in the loop if you really need to use it. Otherwise, I don't see any benefits to introduce this complexity of writing two-level loops.

Reply:
the duplicate logic issue for `get_epoch_idx` we can solve by another utility fn but it's small enough that I didn't. 

hmm one benefit that breaking up the loop will provide would be that it'll let us do operations we need to do on epoch boundaries without having to call context.is_epoch_end every iteration. for example we need it to split per_batch_metrics by epoch for on_train_epoch_end callback. that overhead seems small.

do we support dataloaders where we can't know the length of or with variable lengths?

Reply:
I reverted back to the old loop